morph.go

```diff
@@ -90,36 +94,65 @@ func NewWithDriverAndSource(driver drivers.Driver, source sources.Source, option
 
 // ApplyAll applies all pending migrations.
 func (m *Morph) ApplyAll() error {
+	_, err := m.Apply(-1)
+	return err
+}
+
+// Applies limited number of migrations
+func (m *Morph) Apply(limit int) (int, error) {
 	appliedMigrations, err := m.driver.AppliedMigrations()
 	if err != nil {
-		return err
+		return -1, err
 	}
 
 	pendingMigrations, err := computePendingMigrations(appliedMigrations, m.source.Migrations())
 	if err != nil {
-		return err
+		return -1, err
 	}
 
-	for _, migration := range sortMigrations(pendingMigrations) {
-		start := time.Now()
+	migrations, rollbacks, err := breakDownMigrations(sortMigrations(pendingMigrations), models.Up)
+	if err != nil {
+		return -1, err
+	}
+
+	steps := limit
+	if limit < 0 || len(migrations) < steps {
+		steps = len(migrations)
+	}
 
-		m.config.Logger.Printf(InfoLoggerLight.Sprintf(migrationProgressStart+"\n", migration.Name))
-		if err := m.driver.Apply(migration); err != nil {
-			return err
+	var applied int
+	for i := 0; i < steps; i++ {
+		start := time.Now()
+		migrationName := migrations[i].Name
+		m.config.Logger.Println(InfoLoggerLight.Sprint(formatProgress(fmt.Sprintf(migrationProgressStart, migrationName))))
+		if err := m.driver.Apply(migrations[i], true); err != nil {
+			rollback, ok := rollbacks[migrationName]
```

Comment:
If the `up` migration fails, do we need to apply the `down` equivalent? My assumption would be that the transaction for the `up` migration is not commited and we don't need to do anything but to report the error.